### PR TITLE
Add struct literal support

### PR DESCRIPTION
**Pull request by acquamarin**
Add struct literal support. `Example: return {weight: 5, height: 20}` will return a struct with '{weight: 5, height: 20}'

### CODE REVIEW

File: src/function/vector_list_operation.cpp [hunk 1]
```diff
@@ -140,31 +147,32 @@ std::vector<std::unique_ptr<VectorOperationDefinition>>
...
-    auto bindFunc = [](const std::vector<DataType>& argumentTypes, FunctionDefinition* definition,
-                        DataType& actualReturnType) {
-        if (argumentTypes[0] != argumentTypes[1]) {
-            throw BinderException(getListFunctionIncompatibleChildrenTypeErrorMsg(
-                LIST_CONCAT_FUNC_NAME, argumentTypes[0], argumentTypes[1]));
+    auto bindFunc = [](const binder::expression_vector& argumentTypes,
+                        FunctionDefinition* definition, DataType& actualReturnType) {
+        if (argumentTypes[0]->getDataType() != argumentTypes[1]->getDataType()) {
+            throw BinderException(
+                getListFunctionIncompatibleChildrenTypeErrorMsg(LIST_CONCAT_FUNC_NAME,
```
Comment (andyfengHKU) [inline src/function/vector_list_operation.cpp:154]: should we abstract not only error msg but also the if statement. It seems 152 - 155 is being reused for multiple list bind functions.

File: src/function/vector_struct_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    assert(!parameters.empty() && result.dataType.typeID == common::STRUCT);
+    for (auto& parameter : parameters) {
+        result.addChildVector(parameter);
+    }
+}
+
+void StructPackVectorOperations::structPackBindFunc(const binder::expression_vector& arguments,
+    kuzu::function::FunctionDefinition* definition, common::DataType& actualReturnType) {
+    if (arguments.empty()) {
+        throw common::BinderException("Cannot create an empty struct.");
```
Comment (andyfengHKU) [inline src/function/vector_struct_operations.cpp:18]: check this with test.

File: src/parser/transformer.cpp [hunk 3]
```diff
@@ -756,6 +758,18 @@ std::unique_ptr<ParsedExpression> Transformer::transformListLiteral(
     return listCreation;
 }
 
+std::unique_ptr<ParsedExpression> Transformer::transformStructLiteral(
+    CypherParser::OC_StructLiteralContext& ctx) {
+    auto structPack =
+        std::make_unique<ParsedFunctionExpression>(common::STRUCT_PACK_FUNC_NAME, ctx.getText());
+    for (auto& structField : ctx.oC_StructField()) {
+        auto structExpr = transformExpression(*structField->oC_Expression());
+        structExpr->setAlias(transformSymbolicName(*structField->oC_SymbolicName()));
```
Comment (andyfengHKU) [inline src/parser/transformer.cpp:767]: We probably need to discuss how to access a field from a struct. That part is currently missing and might be tricky to do.

File: src/processor/operator/table_scan/base_table_scan.cpp [hunk 4]
```diff
@@ -6,6 +6,17 @@ namespace processor {
 void BaseTableScan::initLocalStateInternal(ResultSet* resultSet, ExecutionContext* context) {
     for (auto& dataPos : outVecPositions) {
         auto valueVector = resultSet->getValueVector(dataPos);
+        if (valueVector->dataType.getTypeID() == common::STRUCT) {
```
Comment (andyfengHKU) [inline src/processor/operator/table_scan/base_table_scan.cpp:9]: I think this should go into ProcessorTask::populateResultSet

File: src/common/vector/value_vector_utils.cpp [hunk 5]
```diff
@@ -7,19 +7,35 @@ using namespace common;
 
 void ValueVectorUtils::copyNonNullDataWithSameTypeIntoPos(
     ValueVector& resultVector, uint64_t pos, const uint8_t* srcData) {
-    copyNonNullDataWithSameType(resultVector.dataType, srcData,
-        resultVector.getData() + pos * resultVector.getNumBytesPerValue(),
-        resultVector.getOverflowBuffer());
+    if (resultVector.dataType.typeID == STRUCT) {
```
Comment (andyfengHKU) [inline src/common/vector/value_vector_utils.cpp:10]: now that you are changing this. I think we should discuss how LIST can be done in a similar fashion. ideally this copyInto and copyFrom should be handled with

```
switch typeID
recursive types:
   ...
non recursive types:
   ...
```
@ray6080 should think about this as well.

**Review by andyfengHKU — CHANGES_REQUESTED** [review]
This PR is fine but we should against discuss something, specifically struct access and using child vector for list

File: src/antlr4/Cypher.g4 [hunk 6]
```diff
@@ -451,6 +452,12 @@ FALSE : ( 'F' | 'f' ) ( 'A' | 'a' ) ( 'L' | 'l' ) ( 'S' | 's' ) ( 'E' | 'e' ) ;
 oC_ListLiteral
     :  '[' SP? ( oC_Expression SP? ( ',' SP? oC_Expression SP? )* )? ']' ;
 
+oC_StructLiteral
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/antlr4/Cypher.g4:455]: This is not part of the cypher grammar, renamed to `kU_StructLiteral` and `kU_StructField`

File: src/include/parser/expression/parsed_expression.h [hunk 7]
```diff
@@ -33,6 +33,8 @@ class ParsedExpression {
 
     inline ParsedExpression* getChild(uint32_t idx) const { return children[idx].get(); }
 
+    std::vector<ParsedExpression*> getChildren() const;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/parser/expression/parsed_expression.h:36]: removed

File: test/binder/binder_error_test.cpp [hunk 8]
```diff
@@ -502,3 +502,9 @@ TEST_F(BinderErrorTest, MissingStructFields) {
     auto input = "create node table test1(ID INT64, description STRUCT, PRIMARY KEY(ID))";
     ASSERT_STREQ(expectedException.c_str(), getBindingError(input).c_str());
 }
+
+TEST_F(BinderErrorTest, ListCreationOfStruct) {
+    std::string expectedException = "Binder exception: Cannot create a list of structs.";
+    auto input = "RETURN [{a: 5, b: 3}]";
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply test/binder/binder_error_test.cpp:508]: This is valid in duckdb, but we haven't supported this yet. I have a todo in listCreation function.

File: src/function/vector_list_operation.cpp (hunk 1 above)
Comment (acquamarin) [reply src/function/vector_list_operation.cpp:154]: I think the if/else statement is different in each case:
E.g. contains: needs the child type of 1st parameter equals to the second parameter type.
Whereas prepends: needs the type of the 1st parameter equals to the child type of the second parameter.

File: src/processor/operator/table_scan/base_table_scan.cpp (hunk 4 above)
Comment (acquamarin) [reply src/processor/operator/table_scan/base_table_scan.cpp:9]: moved to `ProcessorTask::populateResultSet`

File: src/common/types/types.cpp [hunk 9]
```diff
@@ -1,5 +1,6 @@
 #include "common/types/types.h"
 
+#include <map>
```
Comment (andyfengHKU) [inline src/common/types/types.cpp:3]: Double check if this is actually needed

File: src/processor/processor_task.cpp [hunk 10]
```diff
@@ -38,12 +38,26 @@ std::unique_ptr<ResultSet> ProcessorTask::populateResultSet(
...
+                addStructFieldsVectors(vector.get(), dataChunk.get(), memoryManager);
+            }
             dataChunk->insert(j, std::move(vector));
         }
         resultSet->insert(i, std::move(dataChunk));
     }
     return resultSet;
 }
 
+void ProcessorTask::addStructFieldsVectors(common::ValueVector* structVector,
```
Comment (andyfengHKU) [inline src/processor/processor_task.cpp:51]: Add a TODO for me to move this logic into front-end

**Review by andyfengHKU — APPROVED** [review]

File: src/common/types/types.cpp (hunk 9 above)
Comment (acquamarin) [reply src/common/types/types.cpp:3]: removed

File: src/processor/processor_task.cpp (hunk 10 above)
Comment (acquamarin) [reply src/processor/processor_task.cpp:51]: done

**Pull request merged**